### PR TITLE
Refactor rectangle plot tests - Pass 1

### PR DESCRIPTION
**Pull request by aicioara**
Part of #2636

I will be doing multiple passes because I will get more context (and better sense of best practices) by refactoring other files. However, please flag anything that seems unusual.

As part of this: 
- Applied best practices from the wiki page
- Factored out common code into `beforeEach()` clauses
- No nested `beforeEach()`es
- Correct semantics of the test titles
- Remade hierarchy as shown bellow
screen shot 2015-08-20 at 4 21 49 pm

### CODE REVIEW

**Comment by aicioara** [comment]
[NOQE] As no JS changes

File: test/plots/rectanglePlotTests.ts [hunk 1]
```diff
@@ -2,444 +2,418 @@
...
-    let VERIFY_CELLS = (cells: d3.Selection<any>) => {
-      assert.strictEqual(cells[0].length, 5);
-      cells.each(function(d: any, i: number) {
-        let cell = d3.select(this);
-        assert.closeTo(+cell.attr("height"), 50, 0.5, "Cell height is correct");
-        assert.closeTo(+cell.attr("width"), 50, 0.5, "Cell width is correct");
-        assert.closeTo(+cell.attr("x"), 25 + 50 * i, 0.5, "Cell x coordinate is correct");
-        assert.closeTo(+cell.attr("y"), 25 + 50 * (cells[0].length - i - 1), 0.5, "Cell y coordinate is correct");
+    describe("RectanglePlot - basics", () => {
+      let DATA = [
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: Since we're starting the refactor here...

I don't think `DATA` should be all-capped

File: test/plots/rectanglePlotTests.ts [hunk 2]
```diff
@@ -2,444 +2,418 @@
...
-    ];
-    let VERIFY_CELLS = (cells: d3.Selection<any>) => {
-      assert.strictEqual(cells[0].length, 5);
-      cells.each(function(d: any, i: number) {
-        let cell = d3.select(this);
-        assert.closeTo(+cell.attr("height"), 50, 0.5, "Cell height is correct");
-        assert.closeTo(+cell.attr("width"), 50, 0.5, "Cell width is correct");
-        assert.closeTo(+cell.attr("x"), 25 + 50 * i, 0.5, "Cell x coordinate is correct");
-        assert.closeTo(+cell.attr("y"), 25 + 50 * (cells[0].length - i - 1), 0.5, "Cell y coordinate is correct");
+    describe("RectanglePlot - basics", () => {
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: I feel the "RectanglePlot - " prefix is redundant.

Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: I am really not a fan of the basics description.  It really tells the reader nothing.

File: test/plots/rectanglePlotTests.ts [hunk 3]
```diff
@@ -2,444 +2,418 @@
...
+      let DATA = [
+        { x: 0, y: 0, x2: 1, y2: 1 },
+        { x: 1, y: 1, x2: 2, y2: 2 },
+        { x: 2, y: 2, x2: 3, y2: 3 },
+        { x: 3, y: 3, x2: 4, y2: 4 },
+        { x: 4, y: 4, x2: 5, y2: 5 }
+      ];
+      let VERIFY_CELLS = (cells: d3.Selection<any>) => {
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: Not a fan of the all caps

File: test/plots/rectanglePlotTests.ts [hunk 4]
```diff
@@ -2,444 +2,418 @@
...
-      assert.lengthOf(entities, 1, "found only one entity when querying a point inside the second rectangle");
-      assert.strictEqual(entities[0].index, 1, "entity retrieved is at index 1");
-      entities = plot.entitiesAt({ x: xScale.scale(1), y: yScale.scale(1) });
-      assert.lengthOf(entities, 1, "found only one entity when querying a point inside the first rectangle");
-      assert.strictEqual(entities[0].index, 0, "entity retrieved is at index 0");
-      svg.remove();
-    });
+      afterEach(() => {
+        svg.remove();
+      });
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: We made a conscious decision not to put the `svg.remove()` statement in an `afterEach` block since `afterEach` will run even if the test has failed.  This allows us to quickly up the test on the browser to see what has failed visually without modifying the underlying test.

If there is a way to not run `afterEach()` if the test has failed then that works too.

File: test/plots/rectanglePlotTests.ts [hunk 5]
```diff
@@ -2,444 +2,418 @@
...
-
-      let entities = plot.entitiesIn({
-        topLeft: { x: xScale.scale(1.5), y: yScale.scale(2.5) },
-        bottomRight: { x: xScale.scale(2.5), y: yScale.scale(1.5) } });
-      assert.lengthOf(entities, 2, "retrieved 2 entities intersect with the box");
-      assert.strictEqual(entities[0].index, 1, "the entity of index 1 is retrieved");
-      assert.strictEqual(entities[1].index, 2, "the entity of index 2 is retrieved");
-      svg.remove();
-    });
+      it("renders correctly", () => {
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: This test description is really vague...

File: test/plots/rectanglePlotTests.ts [hunk 6]
```diff
@@ -2,444 +2,418 @@
...
+        assert.lengthOf(entities, 2, "two entities when querying a point in both");
+        assert.strictEqual(entities[0].index, 0, "entity retrieved is at index 0");
+        assert.strictEqual(entities[1].index, 1, "entity retrieved is at index 1");
+        entities = plot.entitiesAt({ x: xScale.scale(4), y: yScale.scale(4) });
+        assert.lengthOf(entities, 1, "found only one entity when querying a point inside the second rectangle");
+        assert.strictEqual(entities[0].index, 1, "entity retrieved is at index 1");
+        entities = plot.entitiesAt({ x: xScale.scale(1), y: yScale.scale(1) });
+        assert.lengthOf(entities, 1, "found only one entity when querying a point inside the first rectangle");
+        assert.strictEqual(entities[0].index, 0, "entity retrieved is at index 0");
+      });
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: Does this test not encapsulate the above test?

File: test/plots/rectanglePlotTests.ts [hunk 7]
```diff
@@ -2,444 +2,418 @@
...
-      plot.x(function(d) { return d.x; }, xScale);
-      plot.y(function(d) { return d.y; }, yScale);
-      plot.y2(function(d) { return d.y2; });
-      plot.addDataset(new Plottable.Dataset(staggeredData));
-      plot.autorangeMode("y");
-      let svg = TestMethods.generateSVG(SVG_WIDTH, SVG_HEIGHT);
-      plot.renderTo(svg);
+      afterEach(() => {
+        svg.remove();
+      });
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: Same comment about `svg.remove()`

File: test/plots/rectanglePlotTests.ts [hunk 8]
```diff
@@ -2,444 +2,418 @@
...
 
-      assert.deepEqual(yScale.domain(), [0, 2], "y domain includes both visible segments");
+      it("autorangeMode(\"x\")", () => {
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: I'm unsure how I feel about the escaping here, it just makes it hard to read.

Would
"autorangeMode on x" be appropriate?

File: test/plots/rectanglePlotTests.ts [hunk 9]
```diff
@@ -2,444 +2,418 @@
...
+        let staggeredData = [
+          { y: "A", x: 0, x2: 1 },
+          { y: "B", x: 1, x2: 2 }
+        ];
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: I like renaming the data variable to be more accurate, but why is this staggered?

File: test/plots/rectanglePlotTests.ts [hunk 10]
```diff
@@ -2,444 +2,418 @@
...
 
-      xScale.domain(["A"]);
-      assert.deepEqual(yScale.domain(), [0, 1], "y domain includes only the visible segment (first)");
+        let xScale = new Plottable.Scales.Linear();
+        let yScale = new Plottable.Scales.Category();
+        xScale.padProportion(0);
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: Just curious on why this needs to be done.

File: test/plots/rectanglePlotTests.ts [hunk 11]
```diff
@@ -2,444 +2,418 @@
...
-        assert.isFalse(Plottable.Utils.Math.isNaN(+sel.attr("height")),
-          "height attribute should be valid for rectangle # " + i + ". Currently " + sel.attr("height"));
-        assert.isFalse(Plottable.Utils.Math.isNaN(+sel.attr("width")),
-          "width attribute should be valid for rectangle # " + i + ". Currently " + sel.attr("width"));
       });
 
-      svg.remove();
-    });
-  });
+      it("autorangeMode(\"y\")", () => {
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: Same comment as above on unit test names

File: test/plots/rectanglePlotTests.ts (hunk 8 above)
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: Also the test name is not aligned with a scenario.  Can we rename the test so that it matches with a wanted scenario?

File: test/plots/rectanglePlotTests.ts [hunk 12]
```diff
@@ -2,444 +2,418 @@
...
-      let gridPlot = new Plottable.Plots.Rectangle();
-      gridPlot.addDataset(dataset)
-              .attr("fill", (d) => d.magnitude, colorScale);
-      gridPlot.x((d: any) => d.x, xScale)
-              .y((d: any) => d.y, yScale)
-              .renderTo(svg);
-      dataset.data(DATA);
-      VERIFY_CELLS((<any> gridPlot)._renderArea.selectAll("rect")[0]);
-      svg.remove();
+    describe("RectanglePlot - fail safe tests", () => {
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: Same comment as above that the prefix here is redundant

File: test/plots/rectanglePlotTests.ts [hunk 13]
```diff
@@ -2,444 +2,418 @@
...
+      it("illegal rectangles don't get displayed", () => {
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: Can you better define what is illegal?

File: test/plots/rectanglePlotTests.ts (hunk 1 above)
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Yup, by doing refactor in other files I've seen it's not even consistent. I'll amend

File: test/plots/rectanglePlotTests.ts (hunk 2 above)
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Agree with point1. What is your suggestion instead of basics? Maybe "general usage".

Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: It would be best if we can narrow down what "general usage" and "basics" mean...

Maybe "renders rectangles if x/y/width/height properties are set"?  I think the reason why our testing infrastructure has slowly decayed is like because we chose ambiguous names in the past

**Comment by bluong** [comment]
Things that I am seeing after this refactor that we may want to consider
* I do not see too much point in having a RectanglePlot prefix for every describe/test.  It is boilerplate and you can see in the shown picture that it is quite obvious that the test is about a RectanglePlot.  This would be more relevant if RectanglePlot was testing behavior with other components... but then that test is no longer a unit test.
* Make the names as precise and as accurate as possible.  I believe poor naming is what got use into a state where we need a refactor due to poor decisions that were justified because of poor naming.  The names have gotten slightly better, but it can definitely be improved.

Less important, but it would help structure our tests
* It has always been interesting to me why we have words like `it` and `describe` but we simply just use them as unit test endpoints and not really semantically.  I am currently not sure how to use the describe block semantically, but the it blocks are almost certainly for things like `it ("informs the user of a success when something good happens")`.  It'd be great if we can structure our tests to be more aligned with our testing software.

File: test/plots/rectanglePlotTests.ts (hunk 6 above)
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Not really. For the above test, the rectangles did not overlap (which is the common case for a rectangle plot). For this one they do. Since this is what the test is testing, I amended the title of the test only and left both tests in

File: test/plots/rectanglePlotTests.ts (hunk 9 above)
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: I like it too, but I don't know what "staggered" means. It is from the previous codebase. Will rename to data for now as it is less ambiguous.

File: test/plots/rectanglePlotTests.ts (hunk 10 above)
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: To avoid computing the padding as well. 

You look at the data and say "oh yea, lowest value is 0, highest one is 2, so I expect the domain to be [0, 2] (instead of [-0.2, 2.2])" Make it as easier for the reader to figure out why is the answer of the assert the one it is.

**Comment by aicioara** [comment]
For your second point, you basically want each test title to begin with a verb. Doable.

**Comment by aicioara** [comment]
Will merge as well as there are some conflicts with `develop`

File: test/plots/rectanglePlotTests.ts [hunk 14]
```diff
@@ -2,511 +2,486 @@
...
+    describe("Basic usage", () => {
+      let data = [
+        { x: 0, y: 0, x2: 1, y2: 1 },
+        { x: 1, y: 1, x2: 2, y2: 2 },
+        { x: 2, y: 2, x2: 3, y2: 3 },
+        { x: 3, y: 3, x2: 4, y2: 4 },
+        { x: 4, y: 4, x2: 5, y2: 5 }
+      ];
+      let verifyCells = (cells: d3.Selection<any>) => {
+        assert.strictEqual(cells[0].length, 5);
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: Shouldn't 5 be `data.length`?  I know this is carryover from the previous implementation, but if we want to do a revamp, let's see what's the best state we can put this test suite in.

File: test/plots/rectanglePlotTests.ts [hunk 15]
```diff
@@ -2,511 +2,486 @@
...
+        cells.each(function(d, i) {
+          let cell = d3.select(this);
+          assert.closeTo(+cell.attr("height"), 50, 0.5, "Cell height is correct");
+          assert.closeTo(+cell.attr("width"), 50, 0.5, "Cell width is correct");
+          assert.closeTo(+cell.attr("x"), 25 + 50 * i, 0.5, "Cell x coordinate is correct");
+          assert.closeTo(+cell.attr("y"), 25 + 50 * (cells[0].length - i - 1), 0.5, "Cell y coordinate is correct");
+        });
+      };
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: Please verify but it seems `verifyCells` is only used to check rectangle widths and heights in a single unit test.  Can we move this there?

File: test/plots/rectanglePlotTests.ts [hunk 16]
```diff
@@ -2,511 +2,486 @@
...
-    });
+      it("retrieves the correct entity under a point", () => {
+        plot.addDataset(new Plottable.Dataset(data));
+        plot.renderTo(svg);
+
+        let entities = plot.entitiesAt({ x: xScale.scale(2.5), y: yScale.scale(2.5) });
+        assert.lengthOf(entities, 1, "found only one entity when querying a point inside the third rectangle");
+        assert.strictEqual(entities[0].index, 2, "entity retrieved is at index 2");
+        svg.remove();
+      });
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: Is this test encapsulated by the following test?

File: test/plots/rectanglePlotTests.ts [hunk 17]
```diff
@@ -2,511 +2,486 @@
...
+        plot.renderTo(svg);
+
+        let entities = plot.entitiesIn(
+          {min: xScale.scale(1.5), max: xScale.scale(2.5)},
+          {min: yScale.scale(2.5), max: yScale.scale(1.5)});
+        assert.lengthOf(entities, 2, "retrieved 2 entities intersect with the box");
+        assert.strictEqual(entities[0].index, 1, "the entity of index 1 is retrieved");
+        assert.strictEqual(entities[1].index, 2, "the entity of index 2 is retrieved");
+        svg.remove();
+      });
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: I would personally wrap the above into a describe block like
`describe("retrieving entities")`

How would that work?

File: test/plots/rectanglePlotTests.ts [hunk 18]
```diff
@@ -2,511 +2,486 @@
...
 
-      let plot = new Plottable.Plots.Rectangle();
-      plot.x(function(d) { return d.x; }, xScale);
-      plot.x2(function(d) { return d.x2; });
-      plot.y(function(d) { return d.y; }, yScale);
-      plot.addDataset(new Plottable.Dataset(staggeredData));
-      plot.autorangeMode("x");
-      let svg = TestMethods.generateSVG(SVG_WIDTH, SVG_HEIGHT);
-      plot.renderTo(svg);
+      it("adjusts the xScale domain with respect to the yScale domain when autorangeMode on x", () => {
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: "when autorangeMode is set to x"?

File: test/plots/rectanglePlotTests.ts [hunk 19]
```diff
@@ -2,511 +2,486 @@
...
-      ];
+        yScale.domain(["B"]);
+        assert.deepEqual(xScale.domain(), [1, 2], "x domain includes only the visible segment (second)");
+        svg.remove();
+      });
 
-      let xScale = new Plottable.Scales.Category();
-      let yScale = new Plottable.Scales.Linear();
-      yScale.padProportion(0);
+      it("adjusts the yScale domain with respect to the xScale domain when autorangeMode on y", () => {
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: Same as above

File: test/plots/rectanglePlotTests.ts [hunk 20]
```diff
@@ -2,511 +2,486 @@
...
+            "height attribute should be valid for rectangle # " + i + ". Currently " + sel.attr("height"));
+          assert.isFalse(Plottable.Utils.Math.isNaN(+sel.attr("width")),
+            "width attribute should be valid for rectangle # " + i + ". Currently " + sel.attr("width"));
+        });
+
+        svg.remove();
+      });
+    });
+
+    describe("Grids", () => {
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: What does "Grids" mean?

File: test/plots/rectanglePlotTests.ts [hunk 21]
```diff
@@ -2,511 +2,486 @@
...
+      let SVG_WIDTH = 400;
+      let SVG_HEIGHT = 200;
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: Let's camelcase this while we're here?

File: test/plots/rectanglePlotTests.ts [hunk 22]
```diff
@@ -2,511 +2,486 @@
...
+        plot.y((d: any) => d.y, yScale);
+        plot.renderTo(svg);
+        dataset.data(data);
+        verifyCells(plot.content().selectAll("rect")[0]);
+        svg.remove();
+      });
+
+      it("renders correctly when there isn't data for every spot", () => {
+        let CELL_HEIGHT = 50;
+        let CELL_WIDTH = 100;
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: more need to camelcase

File: test/plots/rectanglePlotTests.ts [hunk 23]
```diff
@@ -2,511 +2,486 @@
...
+        cellAV = d3.select(cells[2]);
+        cellAU.attr("fill", "#000000");
+        cellAU.attr("x", "0");
+        cellAU.attr("y", "0");
+
+        cellAV.attr("fill", "#ffffff");
+        cellAV.attr("x", "0");
+        cellAV.attr("y", "100");
+        svg.remove();
+      });
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: There are a ton of attribute setters here with 0 assertions being made.  What is this test doing here?

File: test/plots/rectanglePlotTests.ts [hunk 24]
```diff
@@ -2,511 +2,486 @@
...
+        plot.attr("fill", (d) => d.magnitude, colorScale);
+        plot.x((d: any) => d.x, xScale);
+        plot.y((d: any) => d.y, yScale);
+        plot.renderTo(svg);
+        dataset.data(data);
+        verifyCells(plot.content().selectAll("rect")[0]);
+        svg.remove();
+      });
+
+      it("renders correctly when there isn't data for every spot", () => {
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: What does the test name mean?

Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: I see a lot of "renders correctly".  Maybe this should be in a describe block to talk about rendering?

File: test/plots/rectanglePlotTests.ts [hunk 25]
```diff
@@ -2,511 +2,486 @@
...
-        let dataset = new Plottable.Dataset(DATA);
-        gridPlot.addDataset(dataset)
-                .attr("fill", (d) => d.magnitude, colorScale);
-        gridPlot.x((d: any) => d.x, xScale)
-                .y((d: any) => d.y, yScale);
-        gridPlot.renderTo(svg);
-
-        let allCells = gridPlot.selections();
+        let allCells = plot.selections();
         assert.strictEqual(allCells.size(), 4, "all cells retrieved");
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: Do we want to check the contents?

File: test/plots/rectanglePlotTests.ts [hunk 26]
```diff
@@ -2,511 +2,486 @@
...
+        colorScale.range(["black", "white"]);
+
+        plot = new Plottable.Plots.Rectangle<string, string>();
+        plot.attr("fill", (d) => d.magnitude, colorScale);
+        plot.x((d: any) => d.x, xScale);
+        plot.y((d: any) => d.y, yScale);
+
+        dataset = new Plottable.Dataset(data);
+        plot.addDataset(dataset);
+        plot.renderTo(svg);
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: Why do we have this decently specific `beforeEach`?

File: test/plots/rectanglePlotTests.ts [hunk 27]
```diff
@@ -2,511 +2,486 @@
...
-                     .label((d: any) => d.val)
-                     .renderTo(svg);
+        plot = new Plottable.Plots.Rectangle<number, number>();
+        dataset = new Plottable.Dataset(data);
+        plot.addDataset(dataset);
+        plot.x((d: any) => d.x, xScale)
+          .y((d: any) => d.y, yScale)
+          .x2((d: any) => d.x2)
+          .y2((d: any) => d.y2)
+          .label((d: any) => d.val);
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: Why this specific `beforeEach`?

File: test/plots/rectanglePlotTests.ts [hunk 28]
```diff
@@ -2,511 +2,486 @@
...
       });
 
-      it("rectangle labels disabled by default", () => {
+      it("displays rectangle labels by default", () => {
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: does not display?

File: test/plots/rectanglePlotTests.ts [hunk 29]
```diff
@@ -2,511 +2,486 @@
...
+        let svg = TestMethods.generateSVG(150, 300);
+        plot.renderTo(svg);
         let texts = svg.selectAll("text")[0].map((n: any) => d3.select(n).text());
         assert.lengthOf(texts, 0, "by default, no labels are drawn");
         svg.remove();
       });
 
-      it("rectangle labels render properly", () => {
-        rectanglePlot.labelsEnabled(true);
+      it("renders correct number of labels with the correct text", () => {
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: This sounds like two tests?

**Comment by bluong** [comment]
I'm learning how to create a test hierarchy as well as I'm reading this.  I think naming and organization are still the biggest issues here.  While I remember we had a discussion about being careful with our `beforeEach()` usage, there are a few places here where we make assumptions using `beforeEach()`.  Also we should try to use `describe` blocks for better shelving.

File: test/plots/rectanglePlotTests.ts (hunk 29 above)
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Not really. This was supposed to be renders correctly, but I wanted to be more explicit. Whenever we check for the correct text, we should check for the correct number of labels, otherwise we run into the problem of `assert for all (nothing) === true`. My suggestion, I will delete the "correct number of labels bit from the test"

File: test/plots/rectanglePlotTests.ts (hunk 15 above)
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Correct.

File: test/plots/rectanglePlotTests.ts (hunk 16 above)
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Not really. For the above test, the rectangles did not overlap (which is the common case for a rectangle plot). For this one they do. Since this is what the test is testing, I amended the title of the test only and left both tests in.

Commented on this before as well, but the comment got outdated.

File: test/plots/rectanglePlotTests.ts (hunk 21 above)
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Why this change? That is a constant.

File: test/plots/rectanglePlotTests.ts (hunk 23 above)
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Lol, this test was so broken.

File: test/plots/rectanglePlotTests.ts (hunk 26 above)
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Because all this behavior is tested already above. Do you think it is too specific?

File: test/plots/rectanglePlotTests.ts (hunk 27 above)
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: same argument as above

File: test/plots/rectanglePlotTests.ts (hunk 24 above)
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: That would result in duplication. Maybe I'll give it a break until second pass, when I see how I deal with this in other scenarios.

File: test/plots/rectanglePlotTests.ts (hunk 17 above)
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Done

File: test/plots/rectanglePlotTests.ts (hunk 25 above)
Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: Done

File: test/plots/rectanglePlotTests.ts [hunk 30]
```diff
@@ -2,511 +2,519 @@
...
+        let xScale = new Plottable.Scales.Linear();
+        let yScale = new Plottable.Scales.Linear();
+        let plot = new Plottable.Plots.Rectangle<number, number>();
+        plot.x((d) => d.x, xScale).x2((d) => d.x2);
+        plot.y((d) => d.y, yScale).y2((d) => d.y2);
+        plot.addDataset(new Plottable.Dataset(data));
+        plot.renderTo(svg);
+
+        let cells = plot.content().selectAll("rect");
+        assert.strictEqual(cells.size(), data.length);
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: I think we prefer a message here.

File: test/plots/rectanglePlotTests.ts [hunk 31]
```diff
@@ -2,511 +2,519 @@
...
+        cells.each(function(d, i) {
+          let cell = d3.select(this);
+          assert.closeTo(+cell.attr("height"), 50, 0.5, "Cell height is correct");
+          assert.closeTo(+cell.attr("width"), 50, 0.5, "Cell width is correct");
+          assert.closeTo(+cell.attr("x"), 25 + 50 * i, 0.5, "Cell x coordinate is correct");
+          assert.closeTo(+cell.attr("y"), 25 + 50 * (cells[0].length - i - 1), 0.5, "Cell y coordinate is correct");
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: Unsure but I think we want to use window.Pixel_CloseToRequirement here instead of 0.5

File: test/plots/rectanglePlotTests.ts [hunk 32]
```diff
@@ -2,511 +2,519 @@
...
-      let data1 = [
-        { x: "A", y: 1, y2: 2, v: 1 },
-        { x: "B", y: 2, y2: 3, v: 2 },
-        { x: "C", y: 3, y2: NaN, v: 3 },
-        { x: "D", y: 4, y2: 5, v: 4 },
-        { x: "E", y: 5, y2: 6, v: 5 },
-        { x: "F", y: 6, y2: 7, v: 6 }
+    describe("Plots based on Category Scales", () => {
+      let SVG_WIDTH = 400;
+      let SVG_HEIGHT = 200;
```
Comment (bluong) [inline test/plots/rectanglePlotTests.ts]: use `const` instead

Comment (aicioara) [inline test/plots/rectanglePlotTests.ts]: outside the scope of this PR. Will do that once we can enforce it with TSLINT

**Comment by bluong** [comment]
I think this is a huge step up, but probably still quite a bit of work left to go.  Let's do these minor changes and then accumulate more ideas to do later

**Comment by aicioara** [comment]
I will do a second pass anyway.
